### PR TITLE
restructure pvlib/spectrum

### LINKED ISSUE #2125: Proposal: split up mismatch.py
**Is your feature request related to a problem? Please describe.**
I've been working with the spectrum unit of pvlib a fair bit recently for my GSoC project and I find that the [`pvlib/spectrum/mismatch.py`](https://github.com/pvlib/pvlib-python/blob/main/pvlib/spectrum/mismatch.py) file currently contains a range of functions, several of which are not directly related to spectral mismatch.

Current functions in `pvlib/spectrum/mismatch.py`:

[pvlib.spectrum.get_example_spectral_response](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.get_example_spectral_response.html)
[pvlib.spectrum.get_am15g](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.get_am15g.html) (deprecated v0.11, removed 0.12)
[pvlib.spectrum.get_reference_spectra](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.get_reference_spectra.html)
[pvlib.spectrum.calc_spectral_mismatch_field](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.calc_spectral_mismatch_field.html)
[pvlib.spectrum.spectral_factor_caballero](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.spectral_factor_caballero.html)
[pvlib.spectrum.spectral_factor_firstsolar](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.spectral_factor_firstsolar.html)
[pvlib.spectrum.spectral_factor_sapm](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.spectral_factor_sapm.html)
[pvlib.spectrum.spectral_factor_pvspec](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.spectral_factor_pvspec.html)
[pvlib.spectrum.spectral_factor_jrc](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.spectral_factor_jrc.html)
[pvlib.spectrum.sr_to_qe](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.sr_to_qe.html)
[pvlib.spectrum.qe_to_sr](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.spectrum.qe_to_sr.html)

**Describe the solution you'd like**
I am wondering whether it might help organise the functions more clearly if we were to split up this file into several subpackages, still within `pvlib/spectrum`.

One suggestion:

1. `pvlib/spectrum/mismatch.py` 
--> covers the effect on PV performance; calculations of spectral mismatch (atmospheric, field, indoor)

2. `pvlib/spectrum/spectral_irr.py`
--> covers the outdoor meteorological side of things; all spectral irradiance calcs, e.g. `get_reference_spectra`, possibly move `spectrl2` into here as well

4. `pvlib/spectrum/spectral_resp.py` 
--> covers the PV characteristics side of things, e.g. sr_to_qe / qe_to_sr calculations, `get_example_spectral_response`

I think this framework fits with what we have currently and, in terms of future development, I see scope for expansion of `spectral_irr.py` in particular to include for example, function to calculate spectral indices such as the average photon energy. I can see `mismatch.py` being expanded to include variants of the standard spectral mismatch factor, or alternatives such as the (weighted) useful fraction.

**Describe alternatives you've considered**
Could just rename `mismatch.py` entirely to broaden its scope, but I think creating a few files here instead would be better for organisation and future development.

**Additional context**
The suggestion above is just one idea to illustrate what I am thinking about. I am keen to hear other views on this idea.

1. Is splitting `pvlib/spectrum/mismatch.py` a good idea?
If yes:
2. Alternative categorisation structure?
3. Alternative file names?

I'd be happy to work on this if we reach a consensus. I'm looking forward to hearing your views.

### PR DESCRIPTION
**Pull request by RDaxini**
- [X] Closes #2125
 - [X] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [X] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

Context and discussion in #2125

### CODE REVIEW

**Comment by echedey-ls** [comment]
With #2100 a typo was introduced into the list of allowed materials. It is the following:

```
/home/docs/checkouts/readthedocs.org/user_builds/pvlib-python/checkouts/2136/pvlib/spectrum/spectral_mismatch.py:docstring of pvlib.spectrum.spectral_mismatch.spectral_factor_firstsolar:40: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/pvlib-python/checkouts/2136/pvlib/spectrum/spectral_mismatch.py:docstring of pvlib.spectrum.spectral_mismatch.spectral_factor_firstsolar:37: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

Can you address it here? It will only require adding some indent (two spaces) to the `module.` line, no. 159. I can't comment on the moved file since it's not an introduced change.

Sorry for not doing a deep review at that moment.

**Review by echedey-ls — APPROVED** [review]

**Review by cwhanse — APPROVED** [review]

**Comment by RDaxini** [comment]
Thanks @echedey-ls @cwhanse for highlighting these points


> ```
> /home/docs/checkouts/readthedocs.org/user_builds/pvlib-python/checkouts/2136/pvlib/spectrum/spectral_mismatch.py:docstring of pvlib.spectrum.spectral_mismatch.spectral_factor_firstsolar:40: WARNING: Bullet list ends without a blank line; unexpected unindent.
> /home/docs/checkouts/readthedocs.org/user_builds/pvlib-python/checkouts/2136/pvlib/spectrum/spectral_mismatch.py:docstring of pvlib.spectrum.spectral_mismatch.spectral_factor_firstsolar:37: WARNING: Bullet list ends without a blank line; unexpected unindent.
> ```

Done. Where do you check to find these error messages? I remember at the time, when adding the extra ``, I had to create a new line, and wondering how/whether to indent it. Since the main tests passed I figured there was not an issue (I was wrong)


> Can we add a definition of "spectral mismatch"? 

Done. Used your definition. What was there originally was just the original text from the `mismatch.py` file; it was not particularly insightful 😅 I followed a similar style when writing a sentence for the two new files. Let me know if you think any of the other files would benefit from any additional definitions/explanations. Happy to get it done in this same PR.

**Comment by RDaxini** [comment]
Since this is a non-user-facing change, do I still need a whatsnew entry? I think it is a significant change, but not one that affects how users interact with pvlib-python (right?)

**Comment by cwhanse** [comment]
Yes to whatsnew, they also serves as notes for maintainers.

**Comment by RDaxini** [comment]
> also serves as notes for maintainers.

That's good, makes sense. I have added a whatsnew entry to enhancements. Just let me know if there's anything else

**Comment by echedey-ls** [comment]
> Where do you check to find these error messages? 

I wrote a comment recently in which I elaborated on that https://github.com/pvlib/pvlib-python/pull/2124#pullrequestreview-2175571173

You should get to https://readthedocs.org/projects/pvlib-python/builds/ , head to the latest build of your PR, click on the last box with the command `python -m sphinx -T -b html -d _build/doctrees -D language=en . $READTHEDOCS_OUTPUT/html` and find all `warning`, `error`, `critical`. Some of them are persistent across PRs.

> Since the main tests passed I figured there was not an issue 

I hate too that the workflow fails only on critical failures IIRC.

**Comment by RDaxini** [comment]
@echedey-ls thanks for the explanation! And sorry I missed your previous comment:(  Seen it now. That's all very helpful 👍🏽

File: docs/sphinx/source/whatsnew/v0.11.1.rst
Comment (kandersolar) [inline docs/sphinx/source/whatsnew/v0.11.1.rst:23]: These filenames need to be updated

File: pvlib/spectrum/__init__.py
Comment (kandersolar) [inline pvlib/spectrum/__init__.py:8]:
```suggestion
    spectral_factor_jrc,
```

Nitpick: it's nice to retain the trailing comma in a multi-line list like this.  In the future, if we want to add a new element in the list, then the diff is only one line (the new element) rather than two (the new element, plus editing the previous line to add the comma).  See https://www.pythonmorsels.com/trailing-commas-in-python/

File: pvlib/spectrum/irradiance.py
Comment (kandersolar) [inline pvlib/spectrum/irradiance.py:2]: Update filename

File: pvlib/spectrum/response.py
Comment (kandersolar) [inline pvlib/spectrum/response.py:2]: Update filename

**Review by kandersolar — APPROVED** [review]
LGTM after a few easy edits.  Thanks @RDaxini!  I am surprised at how quickly the old `mismatch.py` grew to 1000+ lines of code...

File: pvlib/spectrum/__init__.py
Comment (RDaxini) [reply pvlib/spectrum/__init__.py:8]: Thanks for picking up on that. I have made the change not only to this multiline list but the other two lists fas well.

**Comment by RDaxini** [comment]
> I am surprised at how quickly the old `mismatch.py` grew to 1000+ lines of code...

@kandersolar I guess we have been quite busy lately 😅

I think I have updated everything according to your latest comments now. Let me know if there's anything else.

**Comment by kandersolar** [comment]
Oh, I guess we should also split up `pvlib/tests/test_spectrum.py` to reflect the new structure.  @RDaxini want to add that to this PR, or do a follow-up?

**Comment by RDaxini** [comment]
> Oh, I guess we should also split up `pvlib/tests/test_spectrum.py` to reflect the new structure. 

@kandersolar split up in what way? Do you mean to reorganise the functions within the `test_spectrum.py` file in ordered groups, or create new files such as `test_spectrum_response.py`, `test_spectrum_irradiance.py`?
I think the latter sounds like a decent idea, not sure if I understood correctly though. In this case, would a folder "spectrum" with individual test files be a good idea?

> @RDaxini want to add that to this PR, or do a follow-up?

I can do a follow up PR. There is a lot going on in this one already to reorganise the functions, so a separate one for the tests might help keep the changes in neater/compact chunks.

**Comment by kandersolar** [comment]
> In this case, would a folder "spectrum" with individual test files be a good idea?

Yep, like how the structure of `pvlib/tests/iotools` matches that of `pvlib/iotools`.  a 1:1 correspondence of files, with the test ones having `test_` at the front of the filenames.

**Comment by kandersolar** [comment]
Thanks @RDaxini!

**Pull request merged**
